### PR TITLE
Add `F.scaled_mm`

### PR DESCRIPTION
**Pull request by crcrpar**
## What does this PR do?

As per title, this PR adds `F.scaled_mm` to `thunder.torch` and cover it with torchex impl.

Ref: https://docs.pytorch.org/docs/main/generated/torch.nn.functional.scaled_mm.html

### CODE REVIEW

File: thunder/torch/__init__.py [hunk 1]
```diff
@@ -5990,6 +5990,102 @@ def mse_loss(
...
+            utils.check(
+                len(values) in (1, len(scales)),
+                lambda: (
+                    f"{name} must either be a single value or contain {len(scales)} entries "
+                    f"to match the number of associated scale tensors"
+                ),
+                ValueError,
+            )
+            for enum_value in values:
+                _ = int(enum_value.value) if hasattr(enum_value, "value") else int(enum_value)
```
Comment (mattteochen) [inline thunder/torch/__init__.py:6141]: Should we send a more detailed error here instead of a `TypeError`?

File: thunder/tests/test_ops.py [hunk 2]
```diff
@@ -419,6 +423,394 @@ def fn(a):
...
+    except (NotImplementedError, RuntimeError) as exc:
+        pytest.skip(str(exc))
+
+    jf = thunder.jit(lambda a, b, sa, sb: scaled_mm_fp8(a, b, sa, sb, out_dtype=torch.float32))
+    thunder_out = jf(mat_a_lp, mat_b_lp, decode_a, decode_b)
+
+    emulated = mm_float8_emulated(mat_a_lp, encode_a, mat_b_lp, encode_b, torch.float32)
+
+    assert_close(thunder_out, reference)
+    assert_close(reference, emulated, atol=2e-2, rtol=2e-2)
```
Comment (mattteochen) [inline thunder/tests/test_ops.py]: Given the `assert_closes` CI failures we are seeing lately, maybe it's useful to have a comment explaining why those custom tolerances based on dtype or quantization have been selected (applies to all).

File: thunder/tests/test_ops.py [hunk 3]
```diff
@@ -419,6 +423,394 @@ def fn(a):
...
+
+def _require_fp8_blockwise(device: torch.device) -> None:
+    _require_fp8_rowwise(device)
+
+
+# Adapted from https://github.com/pytorch/pytorch/blob/b4403bfc62ca97eec554cdf815baab1fe93057d9/test/test_scaled_matmul_cuda.py#L645-L659
+@requiresCUDA
+@_require_fp8_tensorwise
+def test_scaled_mm_tensorwise_matches_torch():
+    if not hasattr(torch.nn.functional, "scaled_mm"):
```
Comment (mattteochen) [inline thunder/tests/test_ops.py]: We could make a decorator for this as it's repeated multiple times

File: thunder/torch/__init__.py [hunk 4]
```diff
@@ -5990,6 +5990,102 @@ def mse_loss(
...
+
+    @torchsymbol(torch.nn.functional.scaled_mm)
+    def scaled_mm(
+        mat_a: TensorLike,
+        mat_b: TensorLike,
+        scale_a,
+        scale_recipe_a,
+        scale_b,
+        scale_recipe_b,
+        swizzle_a=None,
```
Comment (kshitij12345) [inline thunder/torch/__init__.py]: It would be helpful to add type annotation for `scale_a`, `scale_recipe_a`, `scale_b`, `scale_recipe_b`, `swizzle_a` and `swizzle_b`.

File: thunder/torch/__init__.py [hunk 5]
```diff
@@ -5990,6 +5990,102 @@ def mse_loss(
...
+            return [t for t in values if isinstance(t, TensorProxy)]
+
+        tensor_args: list[TensorProxy] = [mat_a, mat_b]
+        tensor_args += _tensor_proxy_args(scale_a_list)
+        tensor_args += _tensor_proxy_args(scale_b_list)
+        if isinstance(bias, TensorProxy):
+            tensor_args.append(bias)
+        utils.check_same_device(*tensor_args)
+
+        result_dtype = to_dtype(output_dtype or torch.bfloat16)
```
Comment (kshitij12345) [inline thunder/torch/__init__.py:6158]: Is the `or torch.bfloat16` needed for the case if user explicitly passes `output_dtype=None`?

File: thunder/torch/__init__.py [hunk 6]
```diff
@@ -5990,6 +5990,102 @@ def mse_loss(
...
+            )
+            for enum_value in values:
+                _ = int(enum_value.value) if hasattr(enum_value, "value") else int(enum_value)
+
+        _validate_enum_list(_expand(scale_recipe_a), scale_a_list, "scale_recipe_a")
+        _validate_enum_list(_expand(scale_recipe_b), scale_b_list, "scale_recipe_b")
+        _validate_enum_list(_expand(swizzle_a), scale_a_list, "swizzle_a")
+        _validate_enum_list(_expand(swizzle_b), scale_b_list, "swizzle_b")
+
+        def _tensor_proxy_args(values: Sequence[Any]) -> list[TensorProxy]:
```
Comment (kshitij12345) [inline thunder/torch/__init__.py]: Rename `_tensor_proxy_args` -> `_filter_tensor_proxy_args`.

File: thunder/tests/test_ops.py [hunk 7]
```diff
@@ -419,6 +423,394 @@ def fn(a):
...
+    if (major, minor) < (9, 0):
+        pytest.skip("row-wise scaled_mm requires SM90 or newer")
+    cuda_version = _cuda_version_tuple()
+    if cuda_version is not None and cuda_version < (12, 9):
+        pytest.skip("row-wise scaled_mm requires CUDA 12.9 or newer")
+
+
+def _require_fp8_blockwise(device: torch.device) -> None:
+    _require_fp8_rowwise(device)
+
```
Comment (kshitij12345) [inline thunder/tests/test_ops.py:474]: Some of these helper functions can be moved to thunder/tests/utils.py (especially _cuda_capability and _cuda_version_tuple).

File: thunder/tests/test_ops.py [hunk 8]
```diff
@@ -419,6 +423,394 @@ def fn(a):
...
+
+# Adapted from https://github.com/pytorch/pytorch/blob/b4403bfc62ca97eec554cdf815baab1fe93057d9/test/test_scaled_matmul_cuda.py#L862-L910
+@requiresCUDA
+@_require_fp8_tensorwise
+def test_scaled_mm_matches_emulation():
+    if not hasattr(torch.nn.functional, "scaled_mm"):
+        pytest.skip("torch.nn.functional.scaled_mm is not available in this PyTorch build")
+    device = torch.device("cuda")
+    torch.manual_seed(0)
```
Comment (kshitij12345) [inline thunder/tests/test_ops.py]: It would be better not to set the seed so that we test with different values.

File: thunder/tests/test_ops.py (hunk 2 above)
Comment (kshitij12345) [reply thunder/tests/test_ops.py]: Do we need to test `reference` vs `emulated` given that PyTorch does it? (same question for other cases below)

File: thunder/tests/test_ops.py [hunk 9]
```diff
@@ -419,6 +423,394 @@ def fn(a):
...
+    result = jf(mat_a_lp, mat_b_lp, scale_a, scale_b)
+    assert_close(result, expected)
+
+
+@requiresCUDA
+def test_scaled_mm_rowwise_matches_emulation():
+    if not hasattr(torch.nn.functional, "scaled_mm"):
+        pytest.skip("torch.nn.functional.scaled_mm is not available in this PyTorch build")
+    device = torch.device("cuda")
+    torch.manual_seed(1)
```
Comment (kshitij12345) [inline thunder/tests/test_ops.py]: We shouldn't set seed here. (similar for other cases below)

File: thunder/tests/test_ops.py [hunk 10]
```diff
@@ -419,6 +423,394 @@ def fn(a):
...
+            output_dtype=out_dtype,
+        )
+
+    M, K, N = 32, 64, 32
+    mat_a = torch.randn(M, K, device=device, dtype=torch.float32)
+    mat_b_base = torch.randn(N, K, device=device, dtype=torch.float32)
+
+    mat_a_lp, decode_a, encode_a = quantize_to_fp8(mat_a)
+    mat_b_lp_pre, decode_b, encode_b = quantize_to_fp8(mat_b_base)
+    mat_b_lp = mat_b_lp_pre.t()
```
Comment (crcrpar) [inline thunder/tests/test_ops.py:550]: @kshitij12345 I'm so sorry, I mistakenly deleted your comment instead of mine.

File: thunder/tests/test_ops.py (hunk 7 above)
Comment (t-vi) [reply thunder/tests/test_ops.py:474]: Why do we need `_cuda_capability`? Is something wrong with `torch.cuda.get_device_capability` ?

Comment (crcrpar) [reply thunder/tests/test_ops.py:474]: cleaned it up in the last two commits

File: thunder/torch/__init__.py (hunk 5 above)
Comment (crcrpar) [reply thunder/torch/__init__.py:6158]: I think it's still needed

File: thunder/tests/test_ops.py [hunk 11]
```diff
@@ -419,6 +423,369 @@ def fn(a):
...
+    )
+    thunder_out = fn(mat_a_lp, mat_b_lp, scale_a, scale_b)
+    assert_close(thunder_out, expected)
+
+
+@requiresCUDA
+@_require_scaled_mm
+@pytest.mark.parametrize("output_dtype", [torch.bfloat16])
+@pytest.mark.parametrize("lhs_block,rhs_block", [(1, 1), (128, 1), (1, 128)])
+def test_scaled_mm_blockwise_matches_emulation(output_dtype, lhs_block, rhs_block):
```
Comment (kshitij12345) [inline thunder/tests/test_ops.py]: `test_scaled_mm_blockwise_matches_emulation` seems almost the same as `test_scaled_mm_blockwise_matches_torch` except for the following difference at the end.

```
    reference_f32 = reference.to(torch.float32)
    thunder_out_f32 = thunder_out.to(torch.float32)

    assert_close(thunder_out_f32, reference_f32, atol=7e-1, rtol=7e-2)
```

I think we should keep only one of them.

File: thunder/tests/test_ops.py [hunk 12]
```diff
@@ -419,6 +423,369 @@ def fn(a):
...
+
+    reshaped = tensor.reshape(M // block_rows, block_rows, K // block_cols, block_cols)
+    amax = reshaped.abs().amax(dim=(1, 3), keepdim=True)
+    encode = (max_val / torch.clamp(amax, min=1e-12)).to(torch.float32)
+    quant = torch.clamp(reshaped * encode, min=-max_val, max=max_val).to(dtype_fp8)
+
+    return quant.reshape(M, K), encode.reshape(M // block_rows, K // block_cols).to(tensor.device)
+
+
+def _dequantize_blockwise(quant: torch.Tensor, encode: torch.Tensor, block_rows: int, block_cols: int) -> torch.Tensor:
```
Comment (kshitij12345) [inline thunder/tests/test_ops.py]: This is unused.

File: thunder/tests/test_ops.py [hunk 13]
```diff
@@ -419,6 +423,369 @@ def fn(a):
...
+        pytest.skip(str(exc))
+
+    jf = thunder.jit(reference_fn)
+    result = jf(mat_a_lp, mat_b_lp, scale_a, scale_b)
+    assert_close(result, expected)
+
+
+@requiresCUDA
+@_require_scaled_mm
+def test_scaled_mm_rowwise_matches_emulation():
```
Comment (kshitij12345) [inline thunder/tests/test_ops.py]: IIUC, `test_scaled_mm_rowwise_matches_emulation` actually quantizes the input tensor and generates scale on actual data (unlike `test_scaled_mm_rowwise_matches_torch` which just initializes scales to 1).

`test_scaled_mm_rowwise_matches_emulation` should be renamed to `test_scaled_mm_rowwise_matches_scaled_data` or something instead of `emulation` at the end.

File: thunder/tests/test_ops.py [hunk 14]
```diff
@@ -419,6 +423,369 @@ def fn(a):
...
+    jf = thunder.jit(reference_fn)
+    result = jf(mat_a_lp, mat_b_lp, scale_a, scale_b)
+    assert_close(result, expected)
+
+
+# Adapted from https://github.com/pytorch/pytorch/blob/b4403bfc62ca97eec554cdf815baab1fe93057d9/test/test_scaled_matmul_cuda.py#L862-L910
+@requiresCUDA
+@_require_fp8_tensorwise
+@_require_scaled_mm
+def test_scaled_mm_matches_emulation():
```
Comment (kshitij12345) [inline thunder/tests/test_ops.py]: `test_scaled_mm_matches_emulation` -> `test_scaled_mm_matches_scaled_data` or something. 

Same reasoning as https://github.com/Lightning-AI/lightning-thunder/pull/2720/files#r2527142827

**Comment by crcrpar** [comment]
@KaelanDt could you review this?
